File: ruby/two_sum.rb
# frozen_string_literal: true

# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}

def two_sum(nums, target)
  compliments = {}

  nums.each_with_index do |num, index|
    compliment_index = compliments[num]

    return [compliment_index, index] if compliment_index

    compliments[target - num] = index
  end
end

File: typescript/twoSum.ts
function twoSum(nums: number[], target: number): number[] {
  let compliments: { [key: number]: number } = {};

  for (let i = 0; i < nums.length; i++) {
    let complimentIndex = compliments[nums[i]];

    if (complimentIndex != null) {
      return [complimentIndex, i];
    }

    compliments[target - nums[i]] = i;
  }

  return nums;
}
